### ISSUE TITLE
How is everyone adapting from other languages to java?

### ISSUE THREAD

**Issue by 9frazz** [nus-cs2030/2122-s1#91]
As someone who learned python as my first programming language, I struggled greatly adapting to using java in this module. The simplicity of python really made java a difficult language to learn. How do y'all do it? Any tips or suggestions?

**Comment by RussellDash332** [comment]
Number one is to adapt to how Java needs curly braces and semicolons anywhere and everywhere.
Number two is to compare most of the functions to what is taught on prerequisite modules, like CS1010X for Huzaifa and CS1010S for me.

To those taking CS1010S, stream methods are very similar to that of higher order functions.

**Comment by BrendanCheong** [comment]
Personally starting to find that static typed languages are so much better than dynamically typed languages

**Comment by Ruoming-Lee** [comment]
I learned C language as my first programming language, and then I learned Python in CS1010S. I feel that for beginners Python is quite easy compared to C and Java. And this semester,  it is quite hard for me to catch up with the process of learning Java as I already switched to Python and get rid of the shadow of C.

**Comment by seycara** [comment]
Coming from a Python background (1010X), I kind of understand where the 1010X prof was coming from when we said that Python "isn't really a proper OOP language". I have admittedly got used to accessing attributes as and how I like. However, now that I have learned Java in 2030, I have come to appreciate some of the design choices that Oracle has done for Java in terms of OOP.

**Comment by ryanwalterlee** [comment]
Not too bad as I'm also taking 2040 so pretty much got forced to get used to it faster

**Comment by satsuma42** [comment]
I should preface by stating I'm not a CS kid proper, but life sciences. I started with Python because of CS1010, then picked up R in an LSM-coded mod. Both are very syntactically easy languages, the latter having a lot of leeway with redundancies. Compared to this, Java is an adventure, and not the fun kind. More like the enter-jungle-then-get-malaria-and-die kind. Using vim as an editor is the icing on the cake. 
I will point out, as many have said already, that a statically typed language is a new level of discipline, and so it respecting (or wilfully violating) abstraction when access modifiers are so prevalent and explicit. 
Key observation is that it takes so much more effort to get something working in Java than Python or R. They are clearly made for different things, and as someone who intends to largely handle data and write papers Java is overkill. But I guess it's good practice?

**Comment by Trooksy** [comment]
I agree, python is much more simpler than Java. I had more experience with Javascript, I believe it is more similar than python.
I think Geeksforgeeks really helps out
